libgalois/src/GraphTopology.cpp

```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+    return std::make_unique<ProjectedShuffleTopology>(std::move(pst));
+  }
+
+  // calculate number of new nodes
+  uint32_t num_new_nodes = 0;
+  uint32_t num_new_edges = 0;
+
+  katana::DynamicBitset bitset_nodes;
+  bitset_nodes.resize(topology.num_nodes());
+
+  NUMAArray<Node> old_to_new_nodes_mapping;
+  old_to_new_nodes_mapping.allocateInterleaved(topology.num_nodes());
+
+  katana::ParallelSTL::fill(
+      old_to_new_nodes_mapping.begin(), old_to_new_nodes_mapping.end(),
+      Node{0});
+
+  if (node_types.empty()) {
+    num_new_nodes = topology.num_nodes();
+    // set all nodes
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      bitset_nodes.set(src);
+      old_to_new_nodes_mapping[src] = 1;
+    });
+  } else {
+    std::set<katana::EntityTypeID> node_entity_type_ids;
+
+    for (auto node_type : node_types) {
+      auto entity_type_id = pg->GetNodeEntityTypeID(node_type);
+      node_entity_type_ids.insert(entity_type_id);
+    }
+
+    katana::GAccumulator<uint32_t> accum_num_new_nodes;
+    katana::GAccumulator<uint64_t> accum_num_new_edges;
+
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      for (auto type : node_entity_type_ids) {
+        if (pg->DoesNodeHaveType(src, type)) {
+          accum_num_new_nodes += 1;
+          accum_num_new_edges += std::distance(
+              topology.edges(src).begin(), topology.edges(src).end());
+          bitset_nodes.set(src);
+          old_to_new_nodes_mapping[src] = 1;
+          return;
+        }
+      }
+    });
+    num_new_nodes = accum_num_new_nodes.reduce();
+    num_new_edges = accum_num_new_edges.reduce();
+  }
+
+  // fill old to new nodes mapping
+  katana::ParallelSTL::partial_sum(
+      old_to_new_nodes_mapping.begin(), old_to_new_nodes_mapping.end(),
+      old_to_new_nodes_mapping.begin());
+
+  NUMAArray<Node> new_to_old_nodes_mapping;
+  new_to_old_nodes_mapping.allocateInterleaved(num_new_nodes);
+
+  katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+    if (bitset_nodes.test(src)) {
+      old_to_new_nodes_mapping[src]--;
+      new_to_old_nodes_mapping[old_to_new_nodes_mapping[src]] = src;
+    } else {
+      old_to_new_nodes_mapping[src] = topology.num_nodes();
+    }
+  });
+
+  // calculate number of new edges
+  katana::DynamicBitset bitset_edges;
+  bitset_edges.resize(topology.num_edges());
+
+  NUMAArray<Edge> out_indices;
+  out_indices.allocateInterleaved(num_new_nodes);
+
+  katana::ParallelSTL::fill(out_indices.begin(), out_indices.end(), Edge{0});
+
+  if (edge_types.empty()) {
+    // set all edges incident to projected nodes
+    katana::do_all(
+        katana::iterate(Node{0}, Node{num_new_nodes}),
+        [&](auto src) {
+          auto old_src = new_to_old_nodes_mapping[src];
+          for (Edge e : topology.edges(old_src)) {
+            auto dest = topology.edge_dest(e);
+            if (bitset_nodes.test(dest)) {
+              bitset_edges.set(e);
```

Comment:
Is `DynamicBitset` thread-safe here? Please, comment in the code.

Reply:
those two bitsets are different: one corresponds to nodes and other to edges. And this one: `bitset_edges.set(e); ` is only set by a single thread for a given edge e. thanks

Reply:
Serge, `DynamicBitset` is thread-safe on both CPU and GPU. It supports concurrent reads (test) and concurrent writes (set/reset) but not concurrent reads+writes. This should be documented in the API. If it's not clear, please feel free to edit it and let me know.